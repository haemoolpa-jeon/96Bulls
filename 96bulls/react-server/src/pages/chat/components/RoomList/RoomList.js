import React, { Component } from 'react';
import styles from './RoomList.module.scss';
import firebase from '../../../../config/firebase';
import PropTypes from 'prop-types';
import HashLoader from 'react-spinners/HashLoader';
import { connect } from 'react-redux';
import { setCurrentRoom } from '../../actions';

export class RoomList extends Component {
  state = {
    activeRoom: '',
    room: null,
    user: this.props.currentUser,
    usersRef: firebase.database().ref('users'),
    rooms: [],
    loading: true,
    firstLoad: true,
  };

  componentDidMount() {
    const { user } = this.state;
    if (user) {
      this.addListeners(user.uid);
    }
  }

  /**
   * 컴포넌트가 unmount될 때 이벤트 리스너를 제거합니다.
   */
  componentWillUnmount() {
    this.removeListeners();
  }

  setFirstRoom = () => {
    const firstRoom = this.state.rooms[0];
    if (this.state.firstLoad && this.state.rooms.length > 0) {
      this.props.setCurrentRoom(firstRoom);
      this.setState({ room: firstRoom });
      this.setActiveRoom(firstRoom);
    }
    this.setState({ firstLoad: false });
  };

  setActiveRoom = (room) => {
    this.setState({ activeRoom: room.id });
  };

  changeRoom = (room) => {
    this.setActiveRoom(room);
    this.props.setCurrentRoom(room);
    this.setState({ room });
  };

  addListeners = (userId) => {
    this.addUserListener(userId);
  };

  /**
   * 이벤트 리스너를 등록합니다.
   * DB에서 userId의 방 객체가 추가 될때마다 실행 됩니다.
   * @param userId
   */
  addUserListener = (userId) => {
    const { usersRef } = this.state;
    const loadedRooms = [];
    usersRef.child(userId).child('rooms').on('child_added', (snap) => {
      loadedRooms.push(snap.val());
      this.setState({ rooms: loadedRooms, loading: false }, () => this.setFirstRoom());
    });
  };

  removeListeners = () => {
    this.state.usersRef.off();
  };

  displayRooms = rooms => (
    rooms.length > 0 && rooms.map(room => (
      <button
        type="button"
        onClick={() => this.changeRoom(room)}
        className={`${styles['room-name']} ${this.state.activeRoom === room.id ? styles.active : ''}`}
        key={room.id}
      >
        {`@ ${room.name}`}
      </button>
    ))
  );

  displaySpinner = () => (
    <div className={styles['spinner-wrapper']}>
      <HashLoader color="#ffffff" />
    </div>
  );

  render() {
    const { rooms, loading } = this.state;
    return (
      <section className={styles['room-list']}>
        <h1 className={styles.title}>OPEN CHAT</h1>
        { loading ? this.displaySpinner() : this.displayRooms(rooms)}
      </section>
    );
  }
}

RoomList.propTypes = {
  currentUser: PropTypes.object,
  setCurrentRoom: PropTypes.func.isRequired,
};

export default connect(null, { setCurrentRoom })(RoomList);
