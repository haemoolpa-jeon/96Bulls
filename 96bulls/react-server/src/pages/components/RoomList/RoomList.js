/* Show the current list of rooms that are created */

import React, { Component } from 'react';
import styles from './RoomList.module.scss';
import firebase from '../../../config/firebase';
import PropTypes from 'prop-types';
import HashLoader from 'react-spinners/HashLoader';
import { connect } from 'react-redux';
import { setCurrentRoom } from '../../actions';

export class RoomList extends Component {
  state = {
    activeRoom: '',
    room: null,
    user: this.props.currentUser,
    usersRef: firebase.database().ref('users'),
    rooms: [],
    firstLoad: true,
  };

  componentDidMount() {
    const { user } = this.state;
    if (user) {
      this.addListeners(user.uid);
    }
  }

  
  componentWillUnmount() {
    this.removeListeners();
  }

  setFirstRoom = () => {
    const firstRoom = this.state.rooms[0];
    if (this.state.firstLoad && this.state.rooms.length > 0) {
      this.props.setCurrentRoom(firstRoom);
      this.setState({ room: firstRoom });
      this.setActiveRoom(firstRoom);
    }
    this.setState({ firstLoad: false });
  };

  setActiveRoom = (room) => {
    this.setState({ activeRoom: room.id });
  };

  changeRoom = (room) => {
    this.setActiveRoom(room);
    this.props.setCurrentRoom(room);
    this.setState({ room });
  };

  addListeners = (userId) => {
    this.addUserListener(userId);
  };


  addUserListener = (userId) => {
    const { usersRef } = this.state;
    const loadedRooms = [];
    usersRef.child(userId).child('rooms').on('child_added', (snap) => {
      loadedRooms.push(snap.val());
      this.setState({ rooms: loadedRooms, loading: false }, () => this.setFirstRoom());
    });
  };

  removeListeners = () => {
    this.state.usersRef.off();
  };

  displayRooms = rooms => (
    rooms.length > 0 && rooms.map(room => (
      <button
        type="button"
        onClick={() => this.changeRoom(room)}
        className={`${styles['room-name']} ${this.state.activeRoom === room.id ? styles.active : ''}`}
        key={room.id}
      >
        {`@ ${room.name}`}
      </button>
    ))
  );

  displaySpinner = () => (
    <div className={styles['spinner-wrapper']}>
      <HashLoader color="#ffffff" />
    </div>
  );

  render() {
    const { rooms, loading } = this.state;
    return (
      <section className={styles['room-list']}>
        <h1 className={styles.title}>CLASS CHAT</h1>
        { loading ? this.displaySpinner() : this.displayRooms(rooms)}
      </section>
    );
  }
}

RoomList.propTypes = {
  currentUser: PropTypes.object,
  setCurrentRoom: PropTypes.func.isRequired,
};

export default connect(null, { setCurrentRoom })(RoomList);
