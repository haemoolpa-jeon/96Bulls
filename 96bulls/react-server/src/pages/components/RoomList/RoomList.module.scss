@import "../../styles/variable";

.room-list {
  width: 270px;
  height: 100vh;
  padding: 2rem 0 2rem 0;
  background-color: rgba($color-primary, 0.9);
}

.room-name {
  font-size: 16px;
  color: $color-grey-dark;
  width: 100%;
  padding: 1rem 1rem 1rem 2rem;
  text-align: left;
  &:hover {
    color: $color-grey-light-2;
    cursor: pointer;
    background-color: rgba($color-secondary-dark, .2);
  }
}

.title {
  padding-left: 2rem;
  font-size: 20px;
  color: $color-white;
}

.active {
  background-color: rgba($color-secondary-dark, .4);
}

.spinner-wrapper {
  display: flex;
  justify-content: center;
  margin-top: 10rem;
}
