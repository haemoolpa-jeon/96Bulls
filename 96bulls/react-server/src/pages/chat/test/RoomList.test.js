import React from 'react';
import { RoomList } from '../components/RoomList/RoomList';
import {shallow} from 'enzyme/build';

const mockSetCurrentRoom = jest.fn();

describe('RoomList', () => {
  let component = null;
  it('초기 렌더링이 문제없이 되야함', () => {
    component = shallow(<RoomList setCurrentRoom={mockSetCurrentRoom} />);
  });

  it('초기 렌더링 스냅샷 일치함', () => {
    expect(component).toMatchSnapshot();
  });
});
