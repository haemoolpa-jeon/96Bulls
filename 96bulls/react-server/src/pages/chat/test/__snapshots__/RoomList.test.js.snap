// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomList 초기 렌더링 스냅샷 일치함 1`] = `
<section
  className="room-list"
>
  <h1
    className="title"
  >
    OPEN CHAT
  </h1>
  <div
    className="spinner-wrapper"
  >
    <onlyUpdateForKeys(Loader)
      className=""
      color="#ffffff"
      loading={true}
      size={50}
      sizeUnit="px"
    />
  </div>
</section>
`;
